### ISSUE TITLE
Duplicate Icons On Panel

### ISSUE THREAD

**Issue by mshelby** [lxde/lxqt#1337]
There is an error in either System Tray, lxqt-panel, or CMST that is placing a duplicate networking icon in the system tray, or on the panel. Not sure which, but I will attach a pic that is self-explanatory. The error only occurs with the networking icon which is called through CMST program. The error occurs 90% of the time. Sometimes it is possible to "exit" out of CMST, thereby killing and removing the system tray icon, then restarting CMST which then only shows 1 networking icon... no duplicate.

![lxqt_panel_issue](https://user-images.githubusercontent.com/1563452/29547614-c6fb7ce0-86c0-11e7-8eef-b3d9eda102e5.jpg)

##### Expected Behavior
Any networking application or panel widget should only show one instance of the networking icon. Especially when there is only one network connection utilized. (i.e. both wifi & ethernet are not engaged. Only one or the other is engaged.) 

##### Possible Solution
I would suppose that I am not the only one experiencing this error? In image provided, perhaps the duplication may (or may not!) be caused by offsetting the pixel space allotted to the system tray when it expands due to icons being placed within it? Example: lxqt-panel draws the panel and allots space for the system tray. Meanwhile the users system is starting up networking, connecting to a network and placing an icon in the system tray -which causes the size of the system tray to be expanded... but maybe there's an error in lxqt-panel whereby the panel/system-tray space is not being correctly redrawn? 

The duplicated icon is not clickable & has no functionality, which leads me to believe it has simply been painted by the panel as the system tray expanded. It's really just a ghost image, not an icon. Hope this makes sense. 

##### System Information
LXQT compiled from source on an up-to-date Debian system.

Here's the info pasted from the LXQT-About application:

LXQt Desktop Toolbox - Technical Info<p>
  LXQt About Version:     0.12.2
  LXQt Version:           0.12.0
  Qt:                     5.9.1
  Build type:             Release
  System Configuration:   /etc/xdg
  Share Directory:        /usr/local/share/lxqt
  Translations:           /home/mark/.local/share/lxqt/translations<br>
/usr/local/share/lxqt/translations<br>
/usr/share/lxqt/translations

User Directories
  Xdg Data Home:          /home/mark/.local/share
  Xdg Config Home:        /home/mark/.config
  Xdg Data Dirs:          /home/mark/.local/share:/usr/local/share:/usr/share
  Xdg Cache Home:         /home/mark/.cache
  Xdg Runtime Home:       /run/user/1000
  Xdg Autostart Dirs:     /etc/xdg/autostart
  Xdg Autostart Home:     /home/mark/.config/autostart

**Comment by mshelby** [comment]
Update: I right clicked on system tray and moved it by selected the "move" option. I rearranged the panel so that the system tray is now placed just before the clock widget at the end of the panel. BOTH of the drawn icons were moved along with the system tray so I would assume the issue is either with CMST or with the system-tray.
![lxqt_panel_issue2](https://user-images.githubusercontent.com/1563452/29547929-c4aa822c-86c2-11e7-8c29-9b0426e5ba9d.jpg)

**Comment by tsujan** [comment]
Just a simple test:

(1) Add a temporary panel to the screen bottom;
(2) Remove the Status Notifier plugin from your main panel;
(3) Add the Status Notifier plugin to the temporary panel.

Does the same thing happen?

**Comment by mshelby** [comment]
tsujan, I did as you requested. When I added the status notifier plugin to the newly created temporary panel it showed only one singular icon. The status notifier was not (I think) the offending widget. It displays battery status on my system and is not part of the system tray.

I suppose you know that, of course and wanted me to test another widget, so, no. the test did not result in a duplicate icon being painted on the temporary panel.

Just in case you actually meant for me to delete the system tray widget from my main panel, the recreate it on the temporary panel, I tied that too. Again, only a singular icon was painted on the newly created panel. 

next, I deleted the widget from the temporarily created panel and added the system tray widget back to my original panel. This time it painted only one icon, not two. I will restart my system a few times shortly and report back if the issue repeats itself.

**Comment by tsujan** [comment]
@mshelby  Here, the CMST icon is inside Status Notifier, not the old system tray. I use system tray only for Skype. 2 years ago, when I had Debian, CMST was inside Status Notifier too. I've always had the latest CMST.

**Comment by tsujan** [comment]
Is it so that CMST appears on both status notifier and systray in your case? If so, do you have `sni-qt` installed?

**Comment by mshelby** [comment]
No, I do not have sni-qt installed.
Also, I'm pretty sure that on my lxqt desktop through the Debian system the networking icons have always been placed in the system tray. I built my lxqt desktop completely from source using only the required Debian packages. I used the build script provided by lxqt community and so I don't have any precompiled lxqt packages on my system.

Also, I didn't mean to let on that CMST is putting the icon in both the status notifier AND the system tray. It is placing them only in the system tray.

**Comment by palinek** [comment]
@mshelby how are you starting your lxqt-session? Do you use (are your qt applications) using the `lxqt-qtplugin` (env. variable QT_QPA_PLATFORMTHEME)?

**Comment by mshelby** [comment]
Yes. I am using lxqt-qtplugin. Starting the session via sddm. Totally standard build and install using the build_all.sh script.

**Comment by tsujan** [comment]
For now, the only thing that comes to my mind is this:

(1) Stop the panel from LXQt Session Settings;
(2) Back up and remove (or just rename) `~/.config/lxqt/panel.conf`;
(3) Start the panel again and set it up as before.

If the problem *is* fixed after doing so, there should be a bug in Panel's configuration.

**Comment by mshelby** [comment]
followed your suggested steps. When I started panel again I had one icon. So I logged out and back in. Same old issue. two icons painted on screen.
![lxqt_panel_issue4](https://user-images.githubusercontent.com/1563452/29594668-0d601e0e-8778-11e7-8c74-910fd7bef7fa.jpg)

**Comment by fatdog64** [comment]
I can confirm this problem. Download the wpa_supplicant source and compile wpa_gui. Kill lxqt-panel if it's running, start wpa_gui -t  (-t tells it to start in the tray) and then start lxqt-panel. You'll see two tray icons for wpa_gui. One of them is the real active icon for wpa_gui and the other one is not. We tried modifying wpa_gui's constructor to sleep until QSystemTrayIcon::isSystemTrayAvailable() was true before making the tray icon, but it didn't help, so maybe once the object is created it's too late. 
I made a little program that just sleeps until QSystemTrayIcon::isSystemTrayAvailable() is true. Then in our startup scripts for X apps we call that program so it sleeps until lxqt-panel is up and running, then we start the other apps.  Mshelby, the source for that is here: http://distro.ibiblio.org/fatdog/source/710/trayTest-0.1.tar.xz If you want to compile it you'll need the QT5 headers.  

BTW, Thanks for the great panel!
Kirk

**Comment by mshelby** [comment]
I use firefox browser. I downloaded the 'firetray' addon for it which preloads an instance of the browser and places an icon in the system tray. When icons are sized a good bit smaller than the panel height the panel or tray draws one and a half or two firefox icons too. Hmmm....


Sent from my Samsung Galaxy smartphone.
-------- Original message --------From: Fatdog64 <notifications@github.com> Date: 9/8/17  7:53 PM  (GMT-06:00) To: lxde/lxqt <lxqt@noreply.github.com> Cc: mshelby <mshelby1@gmail.com>, Mention <mention@noreply.github.com> Subject: Re: [lxde/lxqt] Duplicate Icons On Panel (#1337) 
I can confirm this problem. Download the wpa_supplicant source and compile wpa_gui. Kill lxqt-panel if it's running, start wpa_gui -t  (-t tells it to start in the tray) and then start lxqt-panel. You'll see two tray icons for wpa_gui. One of them is the real active icon for wpa_gui and the other one is not. We tried modifying wpa_gui's constructor to sleep until QSystemTrayIcon::isSystemTrayAvailable() was true before making the tray icon, but it didn't help, so maybe once the object is created it's too late.

I made a little program that just sleeps until QSystemTrayIcon::isSystemTrayAvailable() is true. Then in our startup scripts for X apps we call that program so it sleeps until lxqt-panel is up and running, then we start the other apps.  Mshelby, the source for that is here: http://distro.ibiblio.org/fatdog/source/710/trayTest-0.1.tar.xz If you want to compile it you'll need the QT5 headers.
BTW, Thanks for the great panel!

Kirk

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/lxde/lxqt","title":"lxde/lxqt","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/lxde/lxqt"}},"updates":{"snippets":[{"icon":"PERSON","message":"@fatdog64 in #1337: I can confirm this problem. Download the wpa_supplicant source and compile wpa_gui. Kill lxqt-panel if it's running, start wpa_gui -t  (-t tells it to start in the tray) and then start lxqt-panel. You'll see two tray icons for wpa_gui. One of them is the real active icon for wpa_gui and the other one is not. We tried modifying wpa_gui's constructor to sleep until QSystemTrayIcon::isSystemTrayAvailable() was true before making the tray icon, but it didn't help, so maybe once the object is created it's too late. \r\nI made a little program that just sleeps until QSystemTrayIcon::isSystemTrayAvailable() is true. Then in our startup scripts for X apps we call that program so it sleeps until lxqt-panel is up and running, then we start the other apps.  Mshelby, the source for that is here: http://distro.ibiblio.org/fatdog/source/710/trayTest-0.1.tar.xz If you want to compile it you'll need the QT5 headers.  \r\n\r\nBTW, Thanks for the great panel!\r\nKirk\r\n"}],"action":{"name":"View Issue","url":"https://github.com/lxde/lxqt/issues/1337#issuecomment-328243772"}}}

**Comment by tsujan** [comment]
@mshelby

As I mentioned above, CMST should put its icon in the Status Notifier and not system tray. The same is true for wpa_gui, as I tested a minute ago. If you remove System Tray and have Status Notifier on your panel, you should see both icons and only one icon for each app. If their icons appear in System tray instead of Status Notifier -- whether they have single icons or not -- the problem may be outside LXQt.

**Comment by tsujan** [comment]
@fatdog64 

[FeatherNotes](https://github.com/tsujan/FeatherNotes) checks `QSystemTrayIcon::isSystemTrayAvailable()` to create its tray icon and has never had any problem with the status notifier of LXQt panel.

**Comment by fatdog64** [comment]
We do not compile the status notifier plugin for lxqt-panel and I should
also mention that we only use lxqt-panel from the LXQT project. Tsujan, I
can confirm that FeatherNotes does not have this problem. I'm sure if I
spent enough time hacking on wpa_gui I could make it work with lxqt-panel's
system tray, but other programs have problems as well. With Osmo (
http://downloads.sourceforge.net/osmo-pim/) the tray icon doesn't get drawn
correctly, usually it's just a vertical line. Again, this is a problem at
startup when Osmo is running before the system tray is ready and that why
we're now using the program I linked to before to make sure the system tray
is available before starting other X apps.  We also include JWM (Joe's
window manager) which doesn't have these problems with it's system tray,
however JWM is so small and fast perhaps it's tray is ready long before
anything else tries to use it.

Just to be clear, we're quite happy with lxqt-panel, the only reason I
posted was to say that this can be a problem and that it can be worked
around.

**Comment by tsujan** [comment]
> We do not compile the status notifier plugin for lxqt-panel

Why? Status notifier *is* needed by apps with tray icon. It'll be just by chance if FeatherNotes' tray icon works with system tray. As far as I know, system tray is left in LXQt panel only for apps with an old design, like Skype. Apps with an up-to-date code shouldn't have any problem with status notifier.

**Comment by fatdog64** [comment]
Because according to this: https://github.com/lxde/lxqt-panel "Both plugins
are maintained in parallel as not all relevant applications are compatible
with SNI so far. In particular both Qt 4 and all GTK applications need some
kind of wrapper to deal with it." We have many gtk2 apps so I didn't see a
need for both, but perhaps we should give it a go.

**Comment by agaida** [comment]
@fatdog64 - you should, it didn't hurt

**Comment by tsujan** [comment]
> In particular both Qt 4 and all GTK applications need some kind of wrapper to deal with it.

That's right but, in up-to-date systems, you could rarely encounter a Qt4 app.

I don't know about gtk2. The only app with systray icon and requiring gtk that I have is solaar and its icon appears in status notifier.

Status notifier is needed with all up-to-date apps that have tray icons.

**Comment by tsujan** [comment]
@mshelby Do you have status notifier on your panel?

**Comment by mshelby** [comment]
Yes.... I have status notifier on my panel. CMST puts the  internet
connection icon correctly in the status notifier widget. But somehow,
somewhere, someway, through some part of the panel a duplicate icon is
being painted on the panel always directly beside the useable icon.

What would be of interest is to know which sub-module is painting the
unneeded icon. In the pics I've already provided it shows the duplicated
icon, but it doesn't show any other icons in status notifier being
duplicated. That is odd.

Perhaps the internet icon is the last desktop process started (in the
status notifier. If that is true, then it is also the last icon being
painted. Therefore it is always painted at the end of status notifier....
Thus, I can't tell if the status notifier is painting the icon or the panel
(in perhaps a spacing issue?)

Wish I had more apps that could be placed both before and after the
networking icon so that we could KNOW what area it is being painted to.
We've already tried moving the status notifier to it's own panel... It
still duplicates. That still doesn't tell us whether the flaw is in the
panel or the status notifier -since it is again simpy moved to a new panel.

Can anyone tell me a way to surround the networking startup with other
process icons started both before and after networking?


Mark W. Shelby
PH 601.909.0002

**Comment by tsujan** [comment]
> What would be of interest is to know which sub-module is painting the unneeded icon.

Only you could tell us about that by removing either status notifier or systray whenever a double icon is shown. If one of them is inside status notifier and the other inside systray, this issue could be closed, IMO.

**Comment by fatdog64** [comment]
>Status notifier is needed with all up-to-date apps that have tray icons.

I haven't found that to be the case in general, but if you're saying that
lxqt-panel's system tray plugin has problems and the status notifier plugin
does not, then OK. What we have now (making sure lxqt-panel's system tray
is available before running the other X apps) is working well, so I'll
probably leave it at that. Thanks guys, and again thanks to all the LXQT
folks for the nice panel.

**Comment by mshelby** [comment]
Yes. I have done that too. By your reasoning, the error is definitely with
status notifier.

The reason this was confusing at first is because CMST looks specifically
for "system tray" when starting. If it doesn't find it, it throws an error
message. But once it is started, it places it's icon in status notifier....
not in system tray. Placing the icon in status notifier is the correct
procedure.

But I don't think it is correct to simply say the error is solely with
status notifier either.... It could be a spacing issue between panel and
widget? Perhaps panel is holding the spot for status notifier then painting
whatever  last status notifier icon was placed in the held spot. In my
case, networking is the last thing to pre-load, thus it is the duplicated
"ghost: icon.

The only other status notifier icon I have is the laptop battery icon. If I
only had 3 application icons it would be simpler to test just by setting up
a cron script or something to start them in an A-B-C order and force "B" to
be my netwoking process.... If, then the networking icon was still
duplicated we could surmize that the process was either with status
notifier, or with CMST. If one of the other  2 icons was duplicated we
could surmize that the error is squarely with status notifier.

So, I have two pre-loaded icons in status notifier at startup... anybody
know of a third startup process with a status notifier icon that I could
add to my startup? For what it's worth, I don't have bluetooth, so that
won't do...



Mark W. Shelby
PH 601.909.0002

**Comment by tsujan** [comment]
> The reason this was confusing at first is because CMST looks specifically for "system tray" when starting. If it doesn't find it...

That can't be the case. Both on Debian and on Manjaro, I have systray beside status notifier but CMST tray icon has always been correct (inside status notifier).

How do you know the second icon belongs to CMST? Just because the icons are the same? The icon in your screenshots is a generic wireless icon.

**Comment by mshelby** [comment]
@tsujan.... yes, I do. 


Sent from my Samsung Galaxy smartphone.
-------- Original message --------From: tsujan <notifications@github.com> Date: 9/10/17  12:23 PM  (GMT-06:00) To: lxde/lxqt <lxqt@noreply.github.com> Cc: mshelby <mshelby1@gmail.com>, Mention <mention@noreply.github.com> Subject: Re: [lxde/lxqt] Duplicate Icons On Panel (#1337) 
@mshelby Do you have status notifier on your panel?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/lxde/lxqt","title":"lxde/lxqt","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/lxde/lxqt"}},"updates":{"snippets":[{"icon":"PERSON","message":"@tsujan in #1337: @mshelby Do you have status notifier on your panel?"}],"action":{"name":"View Issue","url":"https://github.com/lxde/lxqt/issues/1337#issuecomment-328357684"}}}

**Comment by mshelby** [comment]
I think it would be helpful to go back and re-read this thread from the
start. I've tried most everything you suggest.  The best way I can think of
to test is to isolate both widgets so I have deleted the system tray and
place it on a seperate panel. Next I have placed spacers around the status
notifier. 3 spacers, each side.

Now, I'll wait for it to duplicate the icon once again. If it does not do
so within the next 10 or 12 starts, I would suggest we mark the issue
solved.

If the duplication error occurs once again and it is still placed directly
beside the existing networking icon we can assume the issue is with status
notifier. If the duplication occurs, but it is before or after the three
spacers on either side we can assume the error lies in the panel.

I can't think of a better elimination method to test this out.

Also, no... I do not have any other networking app starting up. I know the
icon is a generic icon... but no. Conman & CMST is the only networking tool
this system utilizes.


Mark W. Shelby
PH 601.909.0002

**Comment by mshelby** [comment]
​


Mark W. Shelby
PH 601.909.0002

**Comment by tsujan** [comment]
@mshelby You didn't tell us how you know both icons belong to CMST. You just replied, "@tsujan.... yes, I do. "

**Comment by mshelby** [comment]
@tsujan At this point i think it might be helpful to go back and read from
the beginning of the thread. also, take a look at the photo's i've
submitted. on several posts i've made certain to state that only one
appears to be a useable icon. The duplicate image appears to simply be
painted in the space on the panel. It is not clickable, hoverable or
interactive in any way.

It seems really simple and obvious to me. I've been very descriptive...
I've tried solutions you offer.... I've provided pictures. I just don't
know any way to state the issue any more clearly for you.

Maybe someone else can be of assistance. As I stated previously (and
included an image!) I've isolated status notifier from the system tray AND
from other widgets on the panel via spacers. I'll let you know if the issue
occurs again. Otherwise, we can close the issue in 10 or 12 days unless
someone else is having the same problem?

Thanks for your willingness to help!


Mark W. Shelby
PH 601.909.0002

**Comment by tsujan** [comment]
> Thanks for your willingness to help!

Sorry, but I had no intention to help you get rid of that icon; I wanted to  know if there was a real bug that none of us could reproduce.

A generic icon can be used by more than one app. Not being clickable or focused only shows that it isn't responsive. So, it is not obvious to me that it belongs to CMST. I just asked how it is obvious to you and got no answer.

**Comment by mshelby** [comment]
.....and thanks for your helpful response. (sarcasm).
I'm sure you are right. I probably imagined the whole thing, then produced
those pictures simply to bait you into replying to this thread so that I
could frustrate you and waste your time. I like Ruth Bell Graham's quote,
"No one is truly useless. If nothing else, at least they can serve as a bad
example."

So, if anyone else has something positive to contribute to the discussion
I'd surely welcome the input. Thanks, mate.


Mark W. Shelby
PH 601.909.0002

**Comment by fatdog64** [comment]
mshelby, does your problem only appear after reboot, X restart, or
logout/login?

**Comment by mshelby** [comment]
@Fatdog64, yes, it usually occurs on any of the boot-up or login scenarios
you mentioned. There seems to be no pattern to it though. Usually it
happens, sometimes it doesn't.

The way to stop the cycle seems to be to kill the CMST process first, (or
select 'exit' in CMST menu. then from the main menu restart 'Conman UI
Setup.'

When the program restarts it usually only paints one icon on the panel...
although sometimes even this method is not foolproof. Sometimes I have to
start and exit the program several times and then shutdown and reboot. In
this, it almost sounds like the image is cached or something, doesn't it?

Its very strange and the issue is not entirely consistent. Sorry I can't be
more specific.


Mark W. Shelby
PH 601.909.0002

**Comment by fatdog64** [comment]
mshelby,

     As I said in my first post, I can duplicate the icon problem using
wpa_gui with the steps I outlined, and I think anyone else could too. I
didn't mention before that this shows up as an intermittent problem in
Fatdog64-710 and only on some computers. It never happened on any of my
computers until I forced it to happen as outlined in my fisrt post. This is
due to Xorg and lxqt-panel taking longer to come up on some computers,
maybe GPU delays. If it takes long enough then some of the X programs get
started before lxqt-panel is ready. So if you know which script launches
CMST you can just put a sleep 3 or maybe 4 before CMST is executed and
you'll probably fix it or you can use the program I linked to. If you can't
find which script runs CMST or if it's ran by some other binary, you could
make a wrapper script for CMST. Rename the CMST binary something like
cmst.bin and then make a script named cmst like this:  (assuming cmst is
the name of the binary)

#!/bin/sh
sleep 4
exec /usr/bin/cmst.bin "$@"

Of course you'll have to change /usr/bin/cmst.bin to whatever that needs to
be and you can adjust the sleep as needed. Also make sure you give the
script the correct permissions. The sleep will give more time for
lxqt-panel to get started.
    It's possible you have some completely different problem, but the
symptoms sound exactly the same.

**Comment by palinek** [comment]
@mshelby can you, please, check if the lxde/lxqt-panel#427 fixes this?

**Comment by fatdog64** [comment]
Your patch fixes it for me with wpa_gui.

**Comment by fatdog64** [comment]
Forgot, Thanks!

**Comment by palinek** [comment]
Great, but we have to thank to @dbungert for the patch!

**Comment by palinek** [comment]
Closing, as this should be fixed by lxde/lxqt-panel#427. If not, please, commend/reopen.
